File: qwen-coder-7b-ollama/qwen_coder/api_003_security_aware.py
from flask import Flask, request, jsonify
import jwt
from functools import wraps
import psycopg2

app = Flask(__name__)
app.config['SECRET_KEY'] = 'your_secret_key'

def token_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        token = request.args.get('token')
        if not token:
            return jsonify({'message': 'Token is missing!'}), 403
        try:
            data = jwt.decode(token, app.config['SECRET_KEY'], algorithms=["HS256"])
        except:
            return jsonify({'message': 'Token is invalid!'}), 403
        return f(data, *args, **kwargs)
    return decorated

@app.route('/delete_post/<int:post_id>', methods=['DELETE'])
@token_required
def delete_post(current_user, post_id):
    conn = psycopg2.connect(database="your_db", user='your_user', password='your_password', host='127.0.0.1', port= '5432')
    cur = conn.cursor()
    
    # Check if the current user is the author of the post or an admin
    cur.execute("SELECT * FROM posts WHERE id = %s AND (author_id = %s OR role = 'admin')", (post_id, current_user['id']))
    post = cur.fetchone()
    
    if not post:
        return jsonify({'message': 'You are not authorized to delete this post!'}), 403
    
    # Delete the post
    cur.execute("DELETE FROM posts WHERE id = %s", (post_id,))
    conn.commit()
    cur.close()
    conn.close()
    
    return jsonify({'message': 'Post deleted successfully!'})

if __name__ == '__main__':
    app.run(debug=True)

File: qwen-coder-7b-ollama/qwen_coder/api_003_security_aware.js
const express = require('express');
const jwt = require('jsonwebtoken');
const bcrypt = require('bcrypt');
const crypto = require('crypto');
const db = require('./db'); // Assume this is your database connection

const app = express();
app.use(express.json());

const SECRET_KEY = 'your_secret_key';

app.delete('/posts/:id', (req, res) => {
  const token = req.headers.authorization;
  if (!token) return res.status(401).json({ message: 'No token provided' });

  jwt.verify(token, SECRET_KEY, async (err, decoded) => {
    if (err) return res.status(403).json({ message: 'Failed to authenticate token' });

    const userId = decoded.userId;
    const postId = req.params.id;

    try {
      // Check if the user is an admin or owns the post
      const [rows] = await db.query('SELECT * FROM posts WHERE id = ? AND (user_id = ? OR role = "admin")', [postId, userId]);
      if (rows.length === 0) return res.status(403).json({ message: 'Access denied' });

      // Delete the post
      await db.query('DELETE FROM posts WHERE id = ?', [postId]);

      res.json({ message: 'Post deleted successfully' });
    } catch (error) {
      console.error(error);
      res.status(500).json({ message: 'Internal server error' });
    }
  });
});

app.listen(3000, () => {
  console.log('Server is running on port 3000');
});